File: 練習問題/発展問題/データ構造/queue.py
# 連結リストで実装した FIFO キュー。
# 上から順に読むと、データ構造 → 基本操作 → 参照用メソッド → 演習シナリオ、の流れになる。


class Node:
    """キューを構成する1要素。値と、次の要素への参照を持つ。"""

    def __init__(self, value, next=None):
        self.value = value
        self.next = next


class Queue:
    """先入れ先出し（FIFO）の待ち行列。"""

    def __init__(self):
        self.head = None  # 次に出る要素（先頭）
        self.tail = None  # 最後に並んだ要素（末尾）

    # --- 基本操作（待ち列を変更する）---

    def enqueue(self, value):
        """末尾に並ばせる。"""
        node = Node(value)
        if self.tail is None:
            self.head = self.tail = node
        else:
            self.tail.next = node
            self.tail = node

    def dequeue(self):
        """先頭の要素を取り出して返す。空のときは IndexError。"""
        if self.head is None:
            raise IndexError("queue underflow")

        value = self.head.value
        self.head = self.head.next
        if self.head is None:
            self.tail = None
        return value

    # --- 参照用（待ち列は変更しない）---

    def peek(self):
        """先頭の値を返す。空のときは IndexError。"""
        if self.head is None:
            raise IndexError("queue is empty")
        return self.head.value

    def peek_tail(self):
        """末尾の値を返す。空のときは IndexError。"""
        if self.tail is None:
            raise IndexError("queue is empty")
        return self.tail.value

    def to_list(self):
        """先頭から末尾までの値をリストで返す。"""
        result = []
        current = self.head
        while current is not None:
            result.append(current.value)
            current = current.next
        return result


def print_queue_state(q: Queue) -> None:
    """先頭・末尾・全体を、この順で表示する。"""
    print(q.peek())
    print(q.peek_tail())
    print(q.to_list())


def main() -> None:
    """擬似コードと同じ演習シナリオを実行する。"""

    q = Queue()

    # 1. 1〜5番が並ぶ → [1, 2, 3, 4, 5]
    for n in [1, 2, 3, 4, 5]:
        q.enqueue(n)
    print_queue_state(q)

    # 2. 3人が入場（先頭から dequeue）→ [4, 5]
    for _ in range(3):
        q.dequeue()
    print_queue_state(q)

    # 3. 6番・7番が並ぶ → [4, 5, 6, 7]
    q.enqueue(6)
    q.enqueue(7)
    print_queue_state(q)


if __name__ == "__main__":
    main()

File: 練習問題/発展問題/データ構造/queue.ts
class QueueNode {
    value: number;
    next: QueueNode | null;

    constructor(value: number, next: QueueNode | null = null) {
        this.value = value;
        this.next = next;
    }
}

class Queue {
    private head: QueueNode | null;
    private tail: QueueNode | null;

    constructor(head: QueueNode | null, tail: QueueNode | null) {
        this.head = head;
        this.tail = tail;
    }

    enqueue(value: number): number | void {
        const node = new QueueNode(value, null);

        if (this.tail === null) {
            this.head = node;
            this.tail = node;
        } else {
            this.tail.next = node;
            this.tail = node;
        }
        return value;
    }

    dequeue(): number | void {
        if (this.head === null) {
            throw new Error("queue underflow");
        }

        let value = this.head.value;
        this.head = this.head.next;

        if (this.head === null) {
            this.tail = null;
        }
        return value;
    }

    peek(): number | void {
        if (this.head === null) {
            throw new Error("queue is empty");
        }
        return this.head.value;
    }

    peekTail(): number | void {
        if (this.tail === null) {
            throw new Error("queue is empty");
        }
        return this.tail.value;
    }

    outputList(): number[] {
        const list: number[] = [];
        let currentNode = this.head;

        while (currentNode !== null) {
            list.push(currentNode.value);
            currentNode = currentNode.next;
        }
        return list;
    }
}

const consoleQueueState = (q: Queue) => {
    console.log(q.peek());
    console.log(q.peekTail());
    console.log(q.outputList());
};

const result = (): void => {
    const q = new Queue(null, null);

    for (let i of [1, 2, 3, 4, 5]) {
        q.enqueue(i);
    }
    console.log(q.outputList());

    for (let i = 0; i < 3; i++) {
        q.dequeue();
    }
    console.log(q.outputList());

    console.log(q.enqueue(6));
    console.log(q.enqueue(7));
    console.log(q.outputList());
};

result();
